<div align="center">

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.png)](https://colab.research.google.com/github/wisaaco/Curs_FG24GOI0074221/blob/main/lessons/2_Pandas.ipynb)


Si no funciona el botó podeu copiar el següent [enllaç](https://colab.research.google.com/github/wisaaco/Curs_FG24GOI0074221/blob/main/lessons/2_Pandas.ipynb)
</div>

# Unitat 2: Llibreria Pandas. Càrrega i emmagatzemament de dades  

Continguts

- 2.A Càrrega selectiva de dades en formats: csv, xlsx i altres formats.
- 2.B Manipulació de dades en Pandas.
- 2.C Visualització
- Activitat

**Pandas** és la llibreria per excel·lència per a l'anàlisi de dades del llenguatge Python. El seu nom prové de “panel data” (terme economètric). Inspirada en les funcionalitats de fulls de càlcul i _R_, però amb el potencial d'aquest llenguatge de propòsit general.

**Pandas** inclou totes les funcionalitats necessàries per al procés d'anàlisi de dades: càrrega, filtrat, tractament, síntesi, agrupament, emmagatzematge i visualització. A més, s'integra amb la resta de llibreries de càlcul numèric com Numpy, visualització amb Matplotlib, aprenentatge automàtic amb scikit-learn, etc.

**L'objectiu d'aquesta unitat és adquirir capacitats bàsiques de càrrega de dades amb Pandas.**

Referències:

- [Pandas](https://pandas.pydata.org/)
- [Python](https://en.wikipedia.org/wiki/Python_(programming_language))
- [Característiques Pandas](https://github.com/pandas-dev/pandas#*main-*features)

*Nota*:  Pandas és una llibrería instal·lada per defecte a l'entorn d'execució de _Google Colab_.

In [3]:
import pandas as pd 

## Càrrega de dades amb  Pandas.

A continuació, obrirem el nostre primer fitxer, ho farem a partir de dades organitzades en una estructura tabular. Aquests fitxers coneguts com a CSV (_Comma Separated Value_) és un dels formats més populars. A Excel pots exportar i importar fitxers en format CSV.

El mètode a Pandas per obrir un fitxer CSV és: [read_csv](https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.read_csv.html).

In [4]:
df_who = pd.read_csv("data/WHO.csv") #Nota: hem d'especificar la ruta (o path) del fitxer

A Pandas hi ha dos tipus **principals** de variables: **dataframes i series**. Aquest mètode genera una variable del tipus dataframe

In [5]:
print(type(df_who))

<class 'pandas.core.frame.DataFrame'>


In [6]:
df_who

Country  CountryID  Continent  Adolescent fertility rate (%)  \
0           Afghanistan          1          1                          151.0   
1               Albania          2          2                           27.0   
2               Algeria          3          3                            6.0   
3               Andorra          4          2                            NaN   
4                Angola          5          3                          146.0   
..                  ...        ...        ...                            ...   
197             Vietnam        198          6                           25.0   
198  West Bank and Gaza        199          1                            NaN   
199               Yemen        200          1                           83.0   
200              Zambia        201          3                          161.0   
201            Zimbabwe        202          3                          101.0   

     Adult literacy rate (%)  \
0                       28.0   
1                       98.7   
2                       69.9   
3                        NaN   
4                       67.4   
..                       ...   
197                     90.3   
198                      NaN   
199                     54.1   
200                     68.0   
201                     89.5   

     Gross national income per capita (PPP international $)  \
0                                                  NaN        
1                                               6000.0        
2                                               5940.0        
3                                                  NaN        
4                                               3890.0        
..                                                 ...        
197                                             2310.0        
198                                                NaN        
199                                             2090.0        
200                                             1140.0        
201                                                NaN        

     Net primary school enrolment ratio female (%)  \
0                                              NaN   
1                                             93.0   
2                                             94.0   
3                                             83.0   
4                                             49.0   
..                                             ...   
197                                           91.0   
198                                            NaN   
199                                           65.0   
200                                           94.0   
201                                           88.0   

     Net primary school enrolment ratio male (%)  \
0                                            NaN   
1                                           94.0   
2                                           96.0   
3                                           83.0   
4                                           51.0   
..                                           ...   
197                                         96.0   
198                                          NaN   
199                                         85.0   
200                                         90.0   
201                                         87.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
0                            26088.0                                4.0  ...   
1                             3172.0                                0.6  ...   
2                            33351.0                                1.5  ...   
3                               74.0                                1.0  ...   
4                            16557.0                                2.8  ...   
..                               ...                                ...  ...   
197                          86206.0                                1.4  ...   

In [7]:
# Estructura
print(df_who.shape) # files o mostres x columnes

(202, 358)


In [8]:
df_who.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Columns: 358 entries, Country to Urban_population_pct_of_total
dtypes: float64(355), int64(2), object(1)
memory usage: 565.1+ KB


In [9]:
df_who.describe()

CountryID   Continent  Adolescent fertility rate (%)  \
count  202.000000  202.000000                     177.000000   
mean   101.500000    3.579208                      59.457627   
std     58.456537    1.808263                      49.105286   
min      1.000000    1.000000                       0.000000   
25%     51.250000    2.000000                      19.000000   
50%    101.500000    3.000000                      46.000000   
75%    151.750000    5.000000                      91.000000   
max    202.000000    7.000000                     199.000000   

       Adult literacy rate (%)  \
count               131.000000   
mean                 78.871756   
std                  20.415760   
min                  23.600000   
25%                  68.400000   
50%                  86.500000   
75%                  95.300000   
max                  99.800000   

       Gross national income per capita (PPP international $)  \
count                                         178.000000        
mean                                        11250.112360        
std                                         12586.753417        
min                                           260.000000        
25%                                          2112.500000        
50%                                          6175.000000        
75%                                         14502.500000        
max                                         60870.000000        

       Net primary school enrolment ratio female (%)  \
count                                     179.000000   
mean                                       84.033520   
std                                        17.788047   
min                                         6.000000   
25%                                        79.000000   
50%                                        90.000000   
75%                                        96.000000   
max                                       100.000000   

       Net primary school enrolment ratio male (%)  \
count                                   179.000000   
mean                                     85.698324   
std                                      15.451212   
min                                      11.000000   
25%                                      79.500000   
50%                                      90.000000   
75%                                      96.000000   
max                                     100.000000   

       Population (in thousands) total  Population annual growth rate (%)  \
count                     1.930000e+02                         193.000000   
mean                      3.409805e+04                           1.297927   
std                       1.304957e+05                           1.163864   
min                       2.000000e+00                          -2.500000   
25%                       1.340000e+03                           0.500000   
50%                       6.762000e+03                           1.300000   
75%                       2.173200e+04                           2.100000   
max                       1.328474e+06                           4.300000   

       Population in urban areas (%)  ...  Total_CO2_emissions  Total_income  \
count                     193.000000  ...         1.860000e+02  1.780000e+02   
mean                       54.911917  ...         1.483596e+05  2.015567e+11   
std                        23.554182  ...         6.133091e+05  9.400689e+11   
min                        10.000000  ...         2.565000e+01  5.190000e+07   
25%                        36.000000  ...         1.672615e+03  3.317500e+09   
50%                        57.000000  ...         1.021157e+04  1.145000e+10   
75%                        73.000000  ...         6.549217e+04  8.680000e+10   
max                       100.000000  ...         5.776432e+06  1.100000e+13   

       Total_reserves  Trade_balance_goods_and_services  \
count      128.000000                      1.710000e+02   
mean        5

In [10]:
df_who.head()

Country  CountryID  Continent  Adolescent fertility rate (%)  \
0  Afghanistan          1          1                          151.0   
1      Albania          2          2                           27.0   
2      Algeria          3          3                            6.0   
3      Andorra          4          2                            NaN   
4       Angola          5          3                          146.0   

   Adult literacy rate (%)  \
0                     28.0   
1                     98.7   
2                     69.9   
3                      NaN   
4                     67.4   

   Gross national income per capita (PPP international $)  \
0                                                NaN        
1                                             6000.0        
2                                             5940.0        
3                                                NaN        
4                                             3890.0        

   Net primary school enrolment ratio female (%)  \
0                                            NaN   
1                                           93.0   
2                                           94.0   
3                                           83.0   
4                                           49.0   

   Net primary school enrolment ratio male (%)  \
0                                          NaN   
1                                         94.0   
2                                         96.0   
3                                         83.0   
4                                         51.0   

   Population (in thousands) total  Population annual growth rate (%)  ...  \
0                          26088.0                                4.0  ...   
1                           3172.0                                0.6  ...   
2                          33351.0                                1.5  ...   
3                             74.0                                1.0  ...   
4                          16557.0                                2.8  ...   

   Total_CO2_emissions  Total_income  Total_reserves  \
0               692.50           NaN             NaN   
1              3499.12  4.790000e+09           78.14   
2            137535.56  6.970000e+10          351.36   
3                  NaN           NaN             NaN   
4              8991.46  1.490000e+10           27.13   

   Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
0                               NaN                         257.00   
1                     -2.040000e+09                          18.47   
2                      4.700000e+09                          40.00   
3                               NaN                            NaN   
4                      9.140000e+09                         164.10   

   Under_five_mortality_from_IHME  Under_five_mortality_rate  \
0                           231.9                     257.00   
1                            15.5                      18.47   
2                            31.2                      40.00   
3                             NaN                        NaN   
4                           242.5                     164.10   

   Urban_population  Urban_population_growth  Urban_population_pct_of_total  
0         5740436.0                     5.44                           22.9  
1         1431793.9                     2.21                           45.4  
2        20800000.0                     2.61                           63.3  
3               NaN                      NaN                            NaN  
4         8578749.0                     4.14                           53.3  

[5 rows x 358 columns]

In [11]:
df_who.tail()

Country  CountryID  Continent  Adolescent fertility rate (%)  \
197             Vietnam        198          6                           25.0   
198  West Bank and Gaza        199          1                            NaN   
199               Yemen        200          1                           83.0   
200              Zambia        201          3                          161.0   
201            Zimbabwe        202          3                          101.0   

     Adult literacy rate (%)  \
197                     90.3   
198                      NaN   
199                     54.1   
200                     68.0   
201                     89.5   

     Gross national income per capita (PPP international $)  \
197                                             2310.0        
198                                                NaN        
199                                             2090.0        
200                                             1140.0        
201                                                NaN        

     Net primary school enrolment ratio female (%)  \
197                                           91.0   
198                                            NaN   
199                                           65.0   
200                                           94.0   
201                                           88.0   

     Net primary school enrolment ratio male (%)  \
197                                         96.0   
198                                          NaN   
199                                         85.0   
200                                         90.0   
201                                         87.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
197                          86206.0                                1.4  ...   
198                              NaN                                NaN  ...   
199                          21732.0                                3.0  ...   
200                          11696.0                                1.9  ...   
201                          13228.0                                0.8  ...   

     Total_CO2_emissions  Total_income  Total_reserves  \
197            101826.23  4.480000e+10           47.11   
198               655.86  3.780000e+09             NaN   
199             20148.34  1.150000e+10          114.52   
200              2366.94  4.090000e+09           10.41   
201             11457.33  5.620000e+09            3.39   

     Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
197                     -1.940000e+09                           20.2   
198                               NaN                           28.0   
199                      8.310000e+08                           82.4   
200                     -4.470000e+08                          175.3   
201                     -1.710000e+08                          106.5   

     Under_five_mortality_from_IHME  Under_five_mortality_rate  \
197                            23.4                       20.2   
198                            25.8                       28.0   
199                            87.9                       82.4   
200                           163.8                      175.3   
201                            67.0                      106.5   

     Urban_population  Urban_population_growth  Urban_population_pct_of_total  
197        21900000.0                     2.90                           26.4  
198         2596216.0                     3.33                           71.6  
199         5759120.5                     4.37                           27.3  
200         4017411.0                     1.95                           35.0  
201         4709965.0                     1.90                           35.9  

[5 rows x 358 columns]

### Carrega de dades (II)

Malauradament, l'estructura i la codificació de les dades en els fitxers _CSV_ varia segons l'eina o el sistema operatiu. Així, podem tenir separadors entre columnes que no siguin la típica ',' o formats no oberts (i.e. utf-8, ansci,...) de codificació del text.

Per això, la funció *read_csv* té molta versatilitat. Mirem la seva [documentació](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html).

Anem a veure que passa quan agafem dades de d'una administració pública: [enllaç](https://datos.gob.es/es/catalogo?q=bilbao&g-recaptcha-response=&administration_level=L&theme_id=economia&sort=score+desc%2C+metadata_created+desc). Aquest fitxer està disponible a: "data/presupuesto_gastos_2023.csv"


In [12]:
df_gastos = pd.read_csv("data/presupuesto_gastos_2023.csv") # quins errors genera ?

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc1 in position 1697: invalid start byte

### Codificació de caràcters

Python utilitza una representació basada en Unicode (https://home.unicode.org/). En canvi altres sistemes operatius i eines opten per altres tipus de representacions.

```python
var = "camión"
var = "lul·lià"
var = "Ζεύς"
var = "ประเทศไทย"
var = "日本語で"
```


Codificacions:

- [Llistat codificacions](https://docs.python.org/3.11/library/codecs.html#standard-encodings)
- [UTF-8](https://es.wikipedia.org/wiki/UTF-8)

In [13]:
# cp1250 | windows-1250 | Central and Eastern Europe

# Mètode bàsic: "Prova i error" !!!
df_gastos = pd.read_csv("data/presupuesto_gastos_2023.csv",delimiter=";",encoding="cp1250")


In [14]:
df_gastos.head()

_id SOZIETATEA_EU/SOCIEDAD_EU SOZIETATEA_CAS/SOCIEDAD_CAS  \
0    1                      UDAL                        UDAL   
1    2                      UDAL                        UDAL   
2    3                      UDAL                        UDAL   
3    4                      UDAL                        UDAL   
4    5                      UDAL                        UDAL   

   EKITALDIA/EJERCICIO SAILA/DEPARTAMENTO  \
0                 2023                110   
1                 2023                110   
2                 2023                110   
3                 2023                110   
4                 2023                110   

  SAILAREN DESKRIBAPENA_EU/DESCRIPCION DEPARTAMENTO_EU  \
0                            KULTURA ETA GOBERNANTZA     
1                            KULTURA ETA GOBERNANTZA     
2                            KULTURA ETA GOBERNANTZA     
3                            KULTURA ETA GOBERNANTZA     
4                            KULTURA ETA GOBERNANTZA     

  SAILAREN DESKRIBAPENA_EU/DESCRIPCION DEPARTAMENTO_CAS  \
0                               CULTURA Y GOBERNANZA      
1                               CULTURA Y GOBERNANZA      
2                               CULTURA Y GOBERNANZA      
3                               CULTURA Y GOBERNANZA      
4                               CULTURA Y GOBERNANZA      

  ZENTRU KUDEATZAILEA/CENTRO GESTOR  \
0                              1100   
1                              1100   
2                              1100   
3                              1100   
4                              1100   

  ZENTRO KUDEATZAILEAREN DESKR._EUS/DESCR. CENTRO GESTOR_EUS  \
0                            KULTURA ETA GOBERNANTZA           
1                            KULTURA ETA GOBERNANTZA           
2                            KULTURA ETA GOBERNANTZA           
3                            KULTURA ETA GOBERNANTZA           
4                            KULTURA ETA GOBERNANTZA           

  ZENTRO KUDEATZAILEAREN DESKR._CAS/DESCR. CENTRO GESTOR_CAS  ...  \
0                               CULTURA Y GOBERNANZA          ...   
1                               CULTURA Y GOBERNANZA          ...   
2                               CULTURA Y GOBERNANZA          ...   
3                               CULTURA Y GOBERNANZA          ...   
4                               CULTURA Y GOBERNANZA          ...   

   ARTIKULUAREN DESKRIBAPENA_CAS/DESCRIPCION ARTICULO_CAS KONTZEPTUA/CONCEPTO  \
0                      RETRIBUCIONES DE ALTOS CARGOS                      100   
1   RETRIBUCIONES DEL PERSONAL EVENTUAL DE GABINETES                      110   
2             RETRIBUCIONES DEL PERSONAL FUNCIONARIO                      120   
3             RETRIBUCIONES DEL PERSONAL FUNCIONARIO                      121   
4             RETRIBUCIONES DEL PERSONAL FUNCIONARIO                      121   

  KONTZEPTUAREN DESKRIBAPENA_EUS/DESCRIPCION CONCEPTO_EUS  \
0  GOI KARGUEN OINARRIZKO SOLDATAK ETA BESTELAKO ...        
1  KABINETEETAKO ALDI BAT. PERTSONALAREN OINARRIZ...        
2                 FUNTZIONARIOEN OINARRIZKO SOLDATAK        
3               FUNTZIONARIOEN ORDAINSARI OSAGARRIAK        
4               FUNTZIONARIOEN ORDAINSARI OSAGARRIAK        

   KONTZEPTUAREN DESKRIBAPENA_CAS/DESCRIPCION CONCEPTO_CAS  \
0  RETRIBUCIONES BASICAS Y OTRAS REMUNERACIONES D...         
1  RETRIBUCIONES BASICAS Y OTRAS REMUNERACIONES D...         
2     RETRIBUCIONES BASICAS DEL PERSONAL FUNCIONARIO         
3  RETRIBUCIONES COMPLEMENTARIAS DEL PERSONAL FUN...         
4  RETRIBUCIONES COMPLEMENTARIAS DEL PERSONAL FUN...         

  AZPIKONTZEPTUA/SUBCONCEPTO  \
0                      10001   
1                      11001   
2                      12001   
3                      12101   
4                      12102   

  AZPIKONTZEPTUAREN DESKRIBAPENA_EUS/DESCRIPCION SUBCONCEPTO_EUS  \
0  GOI KARGUEN OINARRIZKO SOLDATAK ETA BESTELAKO ...               
1  KABINETEETAKO ALDI BAT. PERTSONALAREN OINARRIZ...

### Guardant un dataframe

Durant aquest curs aprendrem a modificar els dataframes, afegirem i eliminarem columnes i també modificarem les que ja tenim. Per tant, feta aquesta feinada, és necessari guardar les dades noves a un fitxer.

In [15]:
df_gastos.to_csv('data/tmp_file.csv',encoding='utf-8') # guardant un dataframe en un fitxer, especificant el format

In [16]:
df_tmp = pd.read_csv("data/tmp_file.csv") # test codificació i separador

### Carrega de dades (III)

Pandas soporta diferents formats de fitxers (i.e. CSV,JSON, formats d'Excel, pàgines web, fitxers penjats a la web, etc.). Obviament, cada format té la seva particularitat.

Trobarem més informació al següent [enllaç](https://pandas.pydata.org/docs/reference/io.html) 

Un exemple de contigut a la Wikipedia:

In [17]:
url = "https://es.wikipedia.org/wiki/Anexo:Comunidades_y_ciudades_aut%C3%B3nomas_de_Espa%C3%B1a" # Fes un copypaste d'aquesta url al teu navigador

comunidades_esp = pd.io.html.read_html(url) 
comunidades_esp # Alerta! Encara no és un dataframe! En una web pots trobar més "taules"

[   Unnamed: 0                      Nombre  \
 0           1                   Andalucía   
 1           2                    Cataluña   
 2           3         Comunidad de Madrid   
 3           4        Comunidad Valenciana   
 4           5                     Galicia   
 5           6             Castilla y León   
 6           7                  País Vasco   
 7           8                    Canarias   
 8           9          Castilla-La Mancha   
 9          10            Región de Murcia   
 10         11                      Aragón   
 11         12              Islas Baleares   
 12         13                 Extremadura   
 13         14      Principado de Asturias   
 14         15  Comunidad Foral de Navarra   
 15         16                   Cantabria   
 16         17                    La Rioja   
 17         18                     Melilla   
 18         19                       Ceuta   
 19      TOTAL                      España   
 
            Capital (de iure o e

In [18]:
print(type(comunidades_esp[0]))
df_comunidades_esp = comunidades_esp[0]

<class 'pandas.core.frame.DataFrame'>


In [19]:
df_comunidades_esp.head() # funciona perfectament??

Unnamed: 0                Nombre Capital (de iure o en su defecto de facto)  \
0          1             Andalucía                                    Sevilla   
1          2              Cataluña                                  Barcelona   
2          3   Comunidad de Madrid                                     Madrid   
3          4  Comunidad Valenciana                                   Valencia   
4          5               Galicia                     Santiago de Compostela   

  Población (2023) Porcentaje población  Densidad (hab./km²) Superficie (km²)  \
0        8 584 147              17,85 %                 9799           87 599   
1        7 901 963              16,43 %                24607           32 113   
2        6 871 903              14,29 %                85599             8028   
3        5 216 195              10,85 %                22430           23 255   
4        2 699 424               5,61 %                 9127           29 575   

  Porcentaje superficie Mapa PIB per cápita en € (2022)  
0               17,31 %  NaN                     21 091  
1                6,35 %  NaN                     32 550  
2                1,59 %  NaN                     38 435  
3                4,60 %  NaN                     24 473  
4                5,84 %  NaN                     25 906

Un segon exemple de contigut amb un fitxer JSON penjat a la web:

In [20]:
#Font: https://data.cityofnewyork.us/Environment/Urban-Park-Ranger-Animal-Condition-Response/fuhs-xmg2
url = 'https://data.cityofnewyork.us/resource/s3vf-x992.json'
df = pd.read_json(url)
df.head()

date_and_time_of_initial  date_and_time_of_ranger        borough  \
0  2021-06-23T16:45:00.000  2021-06-24T08:00:00.000       Brooklyn   
1  2021-06-24T10:00:00.000  2021-06-24T11:00:00.000          Bronx   
2  2021-06-23T14:30:00.000  2021-06-23T14:30:00.000          Bronx   
3  2021-06-23T13:00:00.000  2021-06-23T13:10:00.000  Staten Island   
4  2021-06-23T09:20:00.000  2021-06-23T09:20:00.000         Queens   

                           property  \
0                    Sternberg Park   
1                       Haffen Park   
2                   Pelham Bay Park   
3                  Willowbrook Park   
4  Judge Moses Weinstein Playground   

                                          location species_description  \
0  Inside locked athletic field under construction              Chukar   
1                                      Haffen Pool             Sparrow   
2                                 Pelham Bay South   White-tailed Deer   
3                                     The carousel             Raccoon   
4                                      Garbage can    Virginia Opossum   

  call_source species_status animal_condition  duration_of_response  ...  \
0       Other         Exotic          Healthy                  6.00  ...   
1     Central         Native          Healthy                  1.75  ...   
2    Employee         Native              N/A                  1.00  ...   
3    Employee         Native              N/A                  2.00  ...   
4     Central         Native          Healthy                  2.25  ...   

  _311sr_number final_ranger_action of_animals pep_response  animal_monitored  \
0  311-06712416                 ACC          6        False             False   
1  311-06714879       Rehabilitator          4        False             False   
2           NaN           Unfounded          0        False             False   
3           NaN           Unfounded          0        False             False   
4  311-06699415                 ACC          1        False             False   

   police_response  esu_response  acc_intake_number  hours_spent_monitoring  \
0            False         False             163537                     NaN   
1            False         False                NaN                     NaN   
2            False         False                NaN                     NaN   
3            False         False                NaN                     NaN   
4            False         False             119833                     NaN   

  rehabilitator  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 22 columns]

## <span style="color:red">Activitats</span>

En aquesta activitat practicarem la càrrega de dades en diferents formats. Al món real les dades no sempre tenen una estructura i un format com ens agradaríem.

L'objectiu és que analitzeu la càrrega d'aquestes dades amb les originals: <br/>

- Quina dimensió tenen les dades reals i carregades?
- Quines columnes?
- El concepte de columna com atribut o característica i el concepte de fila com a mostra està present en l'estructura de les dades?
- Coincideixen amb la informació del fitxer?

##### A) "Municipios y fenómeno demográfico de les Illes Balears" <br/>

- Font: https://datos.gob.es/es/catalogo/ea0010587-balears-illes-por-municipios-y-fenomeno-demografico-mnpd-identificador-api-t20-e301-fenom-a2020-l0-23007-px
- Fitxer: "data/municipis.csv" <br/>



In [21]:
import pandas as pd
#TODO

##### B) "Speculation Watch List"
- Font: https://data.cityofnewyork.us/Housing-Development/Speculation-Watch-List/adax-9mit
- Fitxer: "data/Speculation_Watch_List.csv"

In [22]:
import pandas as pd
#TODO

##### C) "Europe Inflation monthly data (annual rate of change)" 

- Font: https://ec.europa.eu/eurostat/databrowser/view/PRC_HICP_MANR__custom_3761882/bookmark/table?lang=en&bookmarkId=4ad27e6f-358a-4a3d-82a0-587d69a833eb
- Fitxer: "data/prc_hicp_manr__custom_3761882_spreadsheet.xlsx" <br/>
- *Nota:*
  - Com podem agafar una "fulla" concreta del full de càlcul? [documentació](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)
  - Per aquesta activitat necessitem instal·lar una llibreria específica: executa la següent cel·la

In [23]:
!pip install openpyxl

In [24]:

import pandas as pd
#TODO


##### D) "Taula de Naixements"

- Font: https://ibestat.caib.es/ibestat/estadistiques/poblacio/naixements/414cab4f-b402-4cd1-af05-6617443de384
- URL dades: "https://ibestat.caib.es/ibestat/estadistiques/poblacio/naixements/414cab4f-b402-4cd1-af05-6617443de384" <br/>



In [25]:
import pandas as pd
#TODO


##### E) "Nights spent at tourist accommodation establishments by residents/non-residents - monthly data"

- Font:  https://ec.europa.eu/eurostat/databrowser/view/tin00171/default/table?lang=en
- Fitxer: "data/tin00171_linear.csv.gz"
- **Nota**: 
    - Els fitxers comprimits amb format .gz podem obrir-se directament com si fossin fitxers de dades amb pandas, i en aquest cas és un fitxer del tipus CSV. És a dir, no cal descomprimir.


In [ ]:
import pandas as pd
#TODO

## Manipulació de dades en Pandas

Ara que ja sabem carregar dataframes de fitxers, descobrirem com podem accedir a l'informació que es troba dins dels tipus de variable fet servir per Pandas.

Un dataframe té columnes i files. Les files són mostres i les columnes representen característiques d'una mostra. Una columna és del tipus Sèrie.

### Selecció d'elements

Pandas ens permet manipular els conjunts de dades tant des d'un de vista de les columnes com de les files. Primer de tot llegim dades com hem vist fins ara.

In [26]:
df_who = pd.read_csv("http://www.exploredata.net/ftp/WHO.csv") #dataframe

In [27]:
df_who.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Columns: 358 entries, Country to Urban_population_pct_of_total
dtypes: float64(355), int64(2), object(1)
memory usage: 565.1+ KB


In [28]:
df_who.describe()

CountryID   Continent  Adolescent fertility rate (%)  \
count  202.000000  202.000000                     177.000000   
mean   101.500000    3.579208                      59.457627   
std     58.456537    1.808263                      49.105286   
min      1.000000    1.000000                       0.000000   
25%     51.250000    2.000000                      19.000000   
50%    101.500000    3.000000                      46.000000   
75%    151.750000    5.000000                      91.000000   
max    202.000000    7.000000                     199.000000   

       Adult literacy rate (%)  \
count               131.000000   
mean                 78.871756   
std                  20.415760   
min                  23.600000   
25%                  68.400000   
50%                  86.500000   
75%                  95.300000   
max                  99.800000   

       Gross national income per capita (PPP international $)  \
count                                         178.000000        
mean                                        11250.112360        
std                                         12586.753417        
min                                           260.000000        
25%                                          2112.500000        
50%                                          6175.000000        
75%                                         14502.500000        
max                                         60870.000000        

       Net primary school enrolment ratio female (%)  \
count                                     179.000000   
mean                                       84.033520   
std                                        17.788047   
min                                         6.000000   
25%                                        79.000000   
50%                                        90.000000   
75%                                        96.000000   
max                                       100.000000   

       Net primary school enrolment ratio male (%)  \
count                                   179.000000   
mean                                     85.698324   
std                                      15.451212   
min                                      11.000000   
25%                                      79.500000   
50%                                      90.000000   
75%                                      96.000000   
max                                     100.000000   

       Population (in thousands) total  Population annual growth rate (%)  \
count                     1.930000e+02                         193.000000   
mean                      3.409805e+04                           1.297927   
std                       1.304957e+05                           1.163864   
min                       2.000000e+00                          -2.500000   
25%                       1.340000e+03                           0.500000   
50%                       6.762000e+03                           1.300000   
75%                       2.173200e+04                           2.100000   
max                       1.328474e+06                           4.300000   

       Population in urban areas (%)  ...  Total_CO2_emissions  Total_income  \
count                     193.000000  ...         1.860000e+02  1.780000e+02   
mean                       54.911917  ...         1.483596e+05  2.015567e+11   
std                        23.554182  ...         6.133091e+05  9.400689e+11   
min                        10.000000  ...         2.565000e+01  5.190000e+07   
25%                        36.000000  ...         1.672615e+03  3.317500e+09   
50%                        57.000000  ...         1.021157e+04  1.145000e+10   
75%                        73.000000  ...         6.549217e+04  8.680000e+10   
max                       100.000000  ...         5.776432e+06  1.100000e+13   

       Total_reserves  Trade_balance_goods_and_services  \
count      128.000000                      1.710000e+02   
mean        5

#### Columnes

Per accedir a la columna podem emprar el nom d'aquesta. El resultat serà un objecte sèrie amb la informació d'aquesta. 

In [29]:
print(df_who.columns)

Index(['Country', 'CountryID', 'Continent', 'Adolescent fertility rate (%)',
       'Adult literacy rate (%)',
       'Gross national income per capita (PPP international $)',
       'Net primary school enrolment ratio female (%)',
       'Net primary school enrolment ratio male (%)',
       'Population (in thousands) total', 'Population annual growth rate (%)',
       ...
       'Total_CO2_emissions', 'Total_income', 'Total_reserves',
       'Trade_balance_goods_and_services', 'Under_five_mortality_from_CME',
       'Under_five_mortality_from_IHME', 'Under_five_mortality_rate',
       'Urban_population', 'Urban_population_growth',
       'Urban_population_pct_of_total'],
      dtype='object', length=358)


Els elements d'una sèrie (columna) els podem gestionar com si fos una llista.

In [30]:
df_who["Country"]

0             Afghanistan
1                 Albania
2                 Algeria
3                 Andorra
4                  Angola
              ...        
197               Vietnam
198    West Bank and Gaza
199                 Yemen
200                Zambia
201              Zimbabwe
Name: Country, Length: 202, dtype: object

In [31]:
print(df_who["Country"][0])
print("-"*30)
print(df_who["Country"][:5]) # slicing
print("-"*30)
print(df_who["Country"].values)

Afghanistan
------------------------------
0    Afghanistan
1        Albania
2        Algeria
3        Andorra
4         Angola
Name: Country, dtype: object
------------------------------
['Afghanistan' 'Albania' 'Algeria' 'Andorra' 'Angola'
 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas' 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus'
 'Belgium' 'Belize' 'Benin' 'Bermuda' 'Bhutan' 'Bolivia'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'Brunei Darussalam'
 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cambodia' 'Cameroon' 'Canada'
 'Cape Verde' 'Central African Republic' 'Chad' 'Chile' 'China' 'Colombia'
 'Comoros' 'Congo, Dem. Rep.' 'Congo, Rep.' 'Cook Islands' 'Costa Rica'
 "Cote d'Ivoire" 'Croatia' 'Cuba' 'Cyprus' 'Czech Republic' 'Denmark'
 'Djibouti' 'Dominica' 'Dominican Republic' 'Ecuador' 'Egypt'
 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Ethiopia' 'Fiji'
 'Finland' 'France' 'French Polynesia' 'Gabon' 'Gambia' 'Georgia'
 'Germany' 'Gh

 Existeix una altra manera més simple de seleccionar una única columna pero hi ha noms de columnes molt llargues: "Children aged &lt;5 years who received any antimalarial treatment for fever (%)"

**Nota**: L'ús de noms de columnes adequats és important per la creació de documents.


In [32]:
print(df_who.columns[9])
print("-"*30)
print(df_who[df_who.columns[9]])

Population annual growth rate (%)
------------------------------
0      4.0
1      0.6
2      1.5
3      1.0
4      2.8
      ... 
197    1.4
198    NaN
199    3.0
200    1.9
201    0.8
Name: Population annual growth rate (%), Length: 202, dtype: float64


#### Files

Cada fila té un índex. L'índex pot ser numèric, alfabètic o de temps.

In [33]:
df_who.index

RangeIndex(start=0, stop=202, step=1)

Emprant el tribut  `loc` del dataframe podem accedir a les seves files amb la mateixa lògica que empram a les llistes:

In [34]:
df_who.loc[0] # la mostra zero, primera fila

Country                           Afghanistan
CountryID                                   1
Continent                                   1
Adolescent fertility rate (%)           151.0
Adult literacy rate (%)                  28.0
                                     ...     
Under_five_mortality_from_IHME          231.9
Under_five_mortality_rate               257.0
Urban_population                    5740436.0
Urban_population_growth                  5.44
Urban_population_pct_of_total            22.9
Name: 0, Length: 358, dtype: object

In [35]:
type(df_who.loc[0]) # una fila també és una sèrie

pandas.core.series.Series

In [36]:
print(df_who.loc[0].Country) # Podem accedir a una sèrie amb el seu índex
print(df_who.loc[0][0]) #o amb la seva posició
print(df_who.loc[0][3])

Afghanistan
Afghanistan
151.0


C:\Users\miquel\AppData\Local\Temp\ipykernel_9960\2098242838.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(df_who.loc[0][0]) #o amb la seva posició
C:\Users\miquel\AppData\Local\Temp\ipykernel_9960\2098242838.py:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(df_who.loc[0][3])


Com passava amb les llistes també podem seleccionar diversos elements (files) en una sola comanda usant la tècnica de _slicing_.

In [37]:
df_who.loc[166:170] #slicing

Country  CountryID  Continent  Adolescent fertility rate (%)  \
166       Somalia        167          3                           35.0   
167  South Africa        168          3                           65.0   
168         Spain        169          2                           10.0   
169     Sri Lanka        170          7                           29.0   
170         Sudan        171          3                           51.0   

     Adult literacy rate (%)  \
166                      NaN   
167                     82.4   
168                     97.2   
169                     90.7   
170                     60.9   

     Gross national income per capita (PPP international $)  \
166                                                NaN        
167                                             8900.0        
168                                            28200.0        
169                                             3730.0        
170                                             1780.0        

     Net primary school enrolment ratio female (%)  \
166                                            6.0   
167                                           88.0   
168                                           99.0   
169                                          100.0   
170                                           37.0   

     Net primary school enrolment ratio male (%)  \
166                                         11.0   
167                                         88.0   
168                                        100.0   
169                                         99.0   
170                                         45.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
166                           8445.0                                3.0  ...   
167                          48282.0                                0.7  ...   
168                          43887.0                                1.1  ...   
169                          19207.0                                0.5  ...   
170                          37707.0                                2.2  ...   

     Total_CO2_emissions  Total_income  Total_reserves  \
166               589.90           NaN             NaN   
167            408792.47  1.610000e+11           66.32   
168            343701.53  6.780000e+11             NaN   
169             11017.65  1.980000e+10           24.28   
170             10618.27  1.670000e+10           10.13   

     Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
166                               NaN                          150.8   
167                     -2.200000e+09                           68.5   
168                     -5.770000e+10                            4.9   
169                     -2.180000e+09                           21.4   
170                     -2.850000e+09                          111.3   

     Under_five_mortality_from_IHME  Under_five_mortality_rate  \
166                           219.9                      150.8   
167                            68.5                       68.5   
168                             4.2                        4.9   
169                            12.1                       21.4   
170                            94.4                      111.3   

     Urban_population  Urban_population_growth  Urban_population_pct_of_total  
166         2885131.0                     4.09                           35.2  
167        27800000.0                     1.98                           59.3  
168        33300000.0                     1.75                           76.7  
169         2969868.0                     0.26                           15.1  
170        15100000.0                     4.40                           40.8  

[5 rows x 358 columns]

L'índex pot canviar-se per qualsevol altre valor (i idealment únic)

In [38]:
df_who.set_index("Country",inplace=True)

print(df_who.index)
print("-"*30)
df_who.loc["Albania"]          # l'accés per LOC ja no és numèric

Index(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia', 'Austria',
       ...
       'United States of America', 'Uruguay', 'Uzbekistan', 'Vanuatu',
       'Venezuela', 'Vietnam', 'West Bank and Gaza', 'Yemen', 'Zambia',
       'Zimbabwe'],
      dtype='object', name='Country', length=202)
------------------------------


CountryID                                                       2.00
Continent                                                       2.00
Adolescent fertility rate (%)                                  27.00
Adult literacy rate (%)                                        98.70
Gross national income per capita (PPP international $)       6000.00
                                                             ...    
Under_five_mortality_from_IHME                                 15.50
Under_five_mortality_rate                                      18.47
Urban_population                                          1431793.90
Urban_population_growth                                         2.21
Urban_population_pct_of_total                                  45.40
Name: Albania, Length: 357, dtype: float64

### Selecció condicional

Si volem que la nostra selecció es correspongui amb un criteri lògic, per exemple saber quins són els països amb una taxa d'alfabetització dels adults amb més d'un 70% podem fer el següent:


In [44]:
df_who = pd.read_csv("http://www.exploredata.net/ftp/WHO.csv") #dataframe
alfabetitzacio = df_who[df_who['Adult literacy rate (%)'] > 70][
    ["Country", "Adult literacy rate (%)"]]
alfabetitzacio

Country  Adult literacy rate (%)
1       Albania                     98.7
6     Argentina                     97.2
7       Armenia                     99.4
10   Azerbaijan                     98.8
12      Bahrain                     86.5
..          ...                      ...
193     Uruguay                     96.8
195     Vanuatu                     75.5
196   Venezuela                     93.0
197     Vietnam                     90.3
201    Zimbabwe                     89.5

[93 rows x 2 columns]

L'estructura d'aquest tipus de selecció sembla complexa, però si la dividim en parts veurem que és abordable:

In [46]:
# En aquest codi obtenim una llista de valors booleans (True o False) segons és compleix la condició per cada una de les files:
seleccio = df_who['Adult literacy rate (%)'] > 70

# En aquest codi, de les files on seleccio == True agafam les dues columnes que ens interessen
df_who[seleccio][["Country","Adult literacy rate (%)"]]

Country  Adult literacy rate (%)
1       Albania                     98.7
6     Argentina                     97.2
7       Armenia                     99.4
10   Azerbaijan                     98.8
12      Bahrain                     86.5
..          ...                      ...
193     Uruguay                     96.8
195     Vanuatu                     75.5
196   Venezuela                     93.0
197     Vietnam                     90.3
201    Zimbabwe                     89.5

[93 rows x 2 columns]

In [47]:
# Multiples criteris
# CO_emisions i Fertilitat
import numpy as np
ix = np.where((df_who["Total_CO2_emissions"] > 10) & (df_who[df_who.columns[3]] <=0.6))
ix

(array([92]),)

I una mostra aleatòria:

In [50]:
import numpy as np
random_sample = np.random.randint(0,2,df_who.shape[0]).astype('bool')
print(random_sample)
print(len(random_sample))

[ True False  True False False False False  True  True  True  True  True
  True False False False False False False  True False  True False False
 False False False False False  True  True False  True  True False  True
 False False False False False  True False  True  True False False False
  True False False  True False  True  True False  True False  True  True
  True  True  True False  True False False False  True False False  True
 False  True  True  True  True  True  True  True  True False False False
 False False  True  True  True  True False False  True  True False False
  True  True False  True  True  True False False  True False  True  True
  True  True False  True  True  True  True False  True  True False  True
 False False  True  True False False False False  True False False False
 False  True  True False  True  True False  True False False False False
  True False  True False False  True False  True  True  True False False
 False  True False  True  True  True False  True Fa

In [51]:
df_who[random_sample]

Country  CountryID  Continent  \
0                 Afghanistan          1          1   
2                     Algeria          3          3   
7                     Armenia          8          2   
8                   Australia          9          6   
9                     Austria         10          2   
..                        ...        ...        ...   
192  United States of America        193          4   
193                   Uruguay        194          5   
194                Uzbekistan        195          2   
195                   Vanuatu        196          6   
199                     Yemen        200          1   

     Adolescent fertility rate (%)  Adult literacy rate (%)  \
0                            151.0                     28.0   
2                              6.0                     69.9   
7                             30.0                     99.4   
8                             16.0                      NaN   
9                             14.0                      NaN   
..                             ...                      ...   
192                           43.0                      NaN   
193                           64.0                     96.8   
194                           40.0                      NaN   
195                           92.0                     75.5   
199                           83.0                     54.1   

     Gross national income per capita (PPP international $)  \
0                                                  NaN        
2                                               5940.0        
7                                               4950.0        
8                                              33940.0        
9                                              36040.0        
..                                                 ...        
192                                            44070.0        
193                                             9940.0        
194                                             2190.0        
195                                             3480.0        
199                                             2090.0        

     Net primary school enrolment ratio female (%)  \
0                                              NaN   
2                                             94.0   
7                                             84.0   
8                                             97.0   
9                                             98.0   
..                                             ...   
192                                           93.0   
193                                          100.0   
194                                           78.0   
195                                           86.0   
199                                           65.0   

     Net primary school enrolment ratio male (%)  \
0                                            NaN   
2                                           96.0   
7                                           80.0   
8                                           96.0   
9                                           97.0   
..                                           ...   
192                                         91.0   
193                                        100.0   
194                                         79.0   
195                                         88.0   
199                                         85.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
0                            26088.0                                4.0  ...   
2                            33351.0                                1.5  ...   
7                             3010.0                               -0.3  ...   
8                            20530.0                                1.1  ...   
9                             8327.0                                0.4  ...   
..                               ...                                ...  ...   
192          

In [52]:
df_who.sample(3) #Nota: hi ha altres mètodes més efectius per fer aquesta funció de sampling


Country  CountryID  Continent  Adolescent fertility rate (%)  \
154    Samoa        155          6                           45.0   
31    Canada         32          4                           14.0   
56   Eritrea         57          3                           85.0   

     Adult literacy rate (%)  \
154                     98.6   
31                       NaN   
56                       NaN   

     Gross national income per capita (PPP international $)  \
154                                             5090.0        
31                                             36280.0        
56                                               680.0        

     Net primary school enrolment ratio female (%)  \
154                                           91.0   
31                                           100.0   
56                                            43.0   

     Net primary school enrolment ratio male (%)  \
154                                         90.0   
31                                          99.0   
56                                          50.0   

     Population (in thousands) total  Population annual growth rate (%)  ...  \
154                            185.0                                0.8  ...   
31                           32577.0                                0.9  ...   
56                            4692.0                                3.6  ...   

     Total_CO2_emissions  Total_income  Total_reserves  \
154               150.22  2.860000e+08           12.46   
31             537523.44  8.100000e+11             NaN   
56                758.45  7.570000e+08            3.80   

     Trade_balance_goods_and_services  Under_five_mortality_from_CME  \
154                     -1.160000e+08                          29.98   
31                       4.270000e+10                           6.10   
56                      -4.020000e+08                          81.80   

     Under_five_mortality_from_IHME  Under_five_mortality_rate  \
154                             NaN                      29.98   
31                              4.9                       6.10   
56                             70.5                      81.80   

     Urban_population  Urban_population_growth  Urban_population_pct_of_total  
154          41181.28                     1.18                           22.4  
31        25900000.00                     1.18                           80.1  
56          878184.06                     5.56                           19.4  

[3 rows x 358 columns]

### Estadístics damunt d'un dataframe

In [54]:
# Creem un dataframe numèric amb valors aleatoris
import numpy as np

np.random.seed(10)

df = pd.DataFrame({"one":np.random.randint(-10,10,5),
                  "two":np.random.random(5),
                  "three":np.random.randint(0,5,5)})
df

one       two  three
0   -1  0.748804      0
1   -6  0.498507      2
2    5  0.224797      0
3  -10  0.198063      4
4    7  0.760531      3

In [55]:
df.sum()

one     -5.000000
two      2.430701
three    9.000000
dtype: float64

In [56]:
df.sum(axis=1) # concepte d'axis 

0    -0.251196
1    -3.501493
2     5.224797
3    -5.801937
4    10.760531
dtype: float64

In [57]:
df.cumsum()

one       two  three
0   -1  0.748804      0
1   -7  1.247311      2
2   -2  1.472108      2
3  -12  1.670170      6
4   -5  2.430701      9

In [58]:
df.apply(np.abs).cumsum()

one       two  three
0    1  0.748804      0
1    7  1.247311      2
2   12  1.472108      2
3   22  1.670170      6
4   29  2.430701      9

In [59]:
df.apply(np.abs).cumsum(axis=1)

one        two      three
0   1.0   1.748804   1.748804
1   6.0   6.498507   8.498507
2   5.0   5.224797   5.224797
3  10.0  10.198063  14.198063
4   7.0   7.760531  10.760531

In [60]:
df.cumprod()

one       two  three
0    -1  0.748804      0
1     6  0.373284      0
2    30  0.083913      0
3  -300  0.016620      0
4 -2100  0.012640      0

In [61]:
df.cumprod(axis=1)

one       two      three
0  -1.0 -0.748804  -0.000000
1  -6.0 -2.991042  -5.982084
2   5.0  1.123983   0.000000
3 -10.0 -1.980629  -7.922515
4   7.0  5.323715  15.971145

In [62]:
ones = np.ones(5)
print(ones)


[1. 1. 1. 1. 1.]


In [63]:
df.sub(ones,axis=0)

one       two  three
0  -2.0 -0.251196   -1.0
1  -7.0 -0.501493    1.0
2   4.0 -0.775203   -1.0
3 -11.0 -0.801937    3.0
4   6.0 -0.239469    2.0

In [64]:
df.sub(ones) #Alerta!

ValueError: Unable to coerce to Series, length must be 3: given 5

In [65]:
df.sub(df.one,axis=0)

one        two  three
0    0   1.748804      1
1    0   6.498507      8
2    0  -4.775203     -5
3    0  10.198063     14
4    0  -6.239469     -4

In [66]:
# nornalització z-score (mean a 0 i desviació a 1)
ts_stand = (df - df.mean()) / df.std()
ts_stand

one       two     three
0  0.000000  0.966021 -1.006231
1 -0.696733  0.045482  0.111803
2  0.836080 -0.961166 -1.006231
3 -1.254119 -1.059487  1.229837
4  1.114773  1.009150  0.670820

In [67]:
print(ts_stand.mean())
print("----")
print(ts_stand.std())

one      0.000000e+00
two      2.664535e-16
three    6.661338e-17
dtype: float64
----
one      1.0
two      1.0
three    1.0
dtype: float64


## <span style="color:red">Activitats</span>

En aquesta activitat practicarem la selecció de dades (columnes i files) en un dataframe

**1) Quina és la mitjana de la població urbana ("Urban_population") de tots els països? La seva desviació típica (std)?**


In [68]:
df_who = pd.read_csv("http://www.exploredata.net/ftp/WHO.csv") #dataframe



**2) Consulta la fila del país: “Malta”**

**3) Quin país té una població urbana més gran?**

**4) El continent on està situat Mala és el mateix que el d'UnitedStates?**

Utilitza una condició per obtenir un resultat Booleà (*True* o *False*)

**5) Quins són els cinc països més contaminants ("Total_CO2_emissions")?**

Aquesta és la meva [pista per a una solució elegant](http://pandas.pydata.org/pandas-docs/version/0.19.2/generated/pandas.DataFrame.sort_values.html)

**6) Observant algunes mostres del fitxer pots establir la relació entre l'identificador del continent i el seu nom?**

És a dir, sabem que Malta és troba continent Europeu i el codi del continent és el 2.

Hi ha els codis de continents: 1, 2, 3, 4, 5, 6, 7

**Nota:** Hi ha dos codis associats a Àsia.

Fes les consultes pertinents al dataframe per construir un diccionari amb la següent

In [71]:
codigoContinentes = {1:"Asia",2:"Europa"}
print(codigoContinentes[2])

Europa


In [72]:
codigoContinentes[3] = ""
codigoContinentes[4] = ""
codigoContinentes[5] = ""
codigoContinentes[6] = ""
codigoContinentes[7] = ""

**6b) Amb el diccionari de codi i nom de continent (activitat 6) transforma aquesta sèrie numèrica de codis en una sèrie més descriptiva que contengui el nom del continent (pista: ``transform()``, ``map()`` o ``apply()``)**


**7) Quins països tenen una població urbana menor a 50.000 ?**

**8) Selecciona els països entre les files 100 i 120 i només 4 columnes a l'atzar.**


**9) Selecciona totes les dades dels països que començen amb la lletra "A".**


Emprant el fitxer ``climaMallorca.csv,`` es demana:<br/>
**10) Quina és la temperatura màxima quan el vent és inferior a 10? Quantes mostres hi ha?**


**11) Quina és la temperatura màxima quan el vent és superior a 10 i inferior a 20? Quantes mostres hi ha?**

## Manipulació de dades en Pandas.

### Modificam columnes 

Tenim diverses maneres d'afegir columnes a un _dataFrame_:

- Mitjançant el nom de la columna que volem afegir, tal com afegim una clau nova a un diccionari.
- `insert`: és un mètode que necessita 3 paràmetres. La posició on volem afegir la columna (`loc`), el seu nom (`column`) i la llista de valors (`value`).
- `assign`: molt semblant a l'anterior, però permet afegir múltiples columnes.

Vegem-ne alguns exemples:

In [74]:
df2 = pd.DataFrame([('Foreign Cinema', 'Restaurant', 289.0),
                    ('Liho Liho', 'Restaurant', 224.0),
                    ('500 Club', 'bar', 80.5),
                    ('The Square', 'bar', 25.30)],
                   columns=('name', 'type', 'AvgBill')
                   )
df2


name        type  AvgBill
0  Foreign Cinema  Restaurant    289.0
1       Liho Liho  Restaurant    224.0
2        500 Club         bar     80.5
3      The Square         bar     25.3

In [75]:
df2['Day'] = "Monday" # Com un diccionari. un ùnic valor a la nova sèrie
df2

name        type  AvgBill     Day
0  Foreign Cinema  Restaurant    289.0  Monday
1       Liho Liho  Restaurant    224.0  Monday
2        500 Club         bar     80.5  Monday
3      The Square         bar     25.3  Monday

In [76]:
df2['Day'] = ['Monday', 'Tuesday', 'Wednesday', 'Thursday']
df2

name        type  AvgBill        Day
0  Foreign Cinema  Restaurant    289.0     Monday
1       Liho Liho  Restaurant    224.0    Tuesday
2        500 Club         bar     80.5  Wednesday
3      The Square         bar     25.3   Thursday

In [77]:
#Emprarem el mètode insert. Indicant on volem inserir aquesta columna amb el paràmetre loc

df2.insert(loc=1, column="Stars", value=[2,2,3,4])
df2

name  Stars        type  AvgBill        Day
0  Foreign Cinema      2  Restaurant    289.0     Monday
1       Liho Liho      2  Restaurant    224.0    Tuesday
2        500 Club      3         bar     80.5  Wednesday
3      The Square      4         bar     25.3   Thursday

El mètode `assing` permet afegir múltiples columnes en una sola operació. És interessant veure que la columna anomenada _AvgHalfBill_ és el resultat de dividir una columna entre 2.

In [78]:
df3 = df2.assign(AvgHalfBill=df2.AvgBill / 2, Michelin_Star=3)
df3

name  Stars        type  AvgBill        Day  AvgHalfBill  \
0  Foreign Cinema      2  Restaurant    289.0     Monday       144.50   
1       Liho Liho      2  Restaurant    224.0    Tuesday       112.00   
2        500 Club      3         bar     80.5  Wednesday        40.25   
3      The Square      4         bar     25.3   Thursday        12.65   

   Michelin_Star  
0              3  
1              3  
2              3  
3              3

### Afegir files

Per afegir files (una nova mostra) a un _dataframe_ tenim el mètode `concat` que concatena dos o més _dataFrames_ separats per comes.

In [80]:
df_files = df2.copy() # Fem una còpia de tota la informació

#Cream un diccionari
fila = {"name":"Isaacs", "type":"bar" ,"AvgBill":[148.2], "day":"Friday"}

nou_df = pd.concat([df_files, pd.DataFrame(fila)])

### Eliminar columnes i files
Molts cops tendrem la necessitat d'eliminar informació que no ens és interessant.

Tenim el mètode `drop` que ens proporciona un nou _dataFrame_ sense la(es) fila(es) o la(les) columna(es) que seleccionem.

Si volem eliminar columnes podem fer-ho especificant la llista de columnes al paràmetre `columns` de la següent manera:

In [81]:
df_no_michelin = df3.drop(columns=["Michelin_Star"]) # Eliminam la darrera columna que hem afegit
df_no_michelin

name  Stars        type  AvgBill        Day  AvgHalfBill
0  Foreign Cinema      2  Restaurant    289.0     Monday       144.50
1       Liho Liho      2  Restaurant    224.0    Tuesday       112.00
2        500 Club      3         bar     80.5  Wednesday        40.25
3      The Square      4         bar     25.3   Thursday        12.65

Per poder eliminar files, fem servir la mateixa funció `drop`, aquesta vegada sense el paràmetre que hem fet servir anteriorment, simplement indiquem els índexs a eliminar:

In [82]:
df_less_rows = df_no_michelin.drop([1,3])
df_less_rows

name  Stars        type  AvgBill        Day  AvgHalfBill
0  Foreign Cinema      2  Restaurant    289.0     Monday       144.50
2        500 Club      3         bar     80.5  Wednesday        40.25

### Concatenació i unió de dataframes

A vegades les dades venem en diferents fitxers i necessiten un únic fitxer, aquest procés implica una concatenació de dataframes. I en altres ocasions, les dades són complementàries és a dir, hi ha noves columnes d'un dataframe i això es coneix com fer operacions d'unió (joins).

**Concatenació**

In [ ]:
df1 = pd.DataFrame([('Foreign Cinema', 'Restaurant', 289.0),
                   ('Liho Liho', 'Restaurant', 224.0),
                   ('500 Club', 'bar', 80.5),
                   ('The Square', 'bar', 25.30)],
           columns=('name', 'type', 'AvgBill')
                 )
df1

In [ ]:
df2 = pd.DataFrame([('Biels', 'Quiosquet', 389.0),
                   ('Barkus', 'Bar', 24.0),
                   ('Blue Wall', 'bar', 80.5),
                   ('Bounty Hunters', 'Social Club', 125.30)],
           columns=('name', 'type', 'AvgBill')
                 )
df2

In [ ]:
dfAll = pd.concat([df1,df2])
dfAll

In [ ]:
dfAll = dfAll.reset_index()
dfAll

In [ ]:
dfAll.drop(columns=["index"])

In [ ]:
# https://pandas.pydata.org/docs/reference/api/pandas.concat.html
dfAll = pd.concat([df1,df2],ignore_index=True)
dfAll

**Unió**

Més informació: https://pandas.pydata.org/docs/user_guide/merging.html

In [ ]:
df1 = pd.DataFrame([('Jhon', 1, "Italy"),
                   ('Pep', 2, "Germany"),
                   ('William', 3, "Finland"),
                   ('Snake', 4, "Italy")],
           columns=('name', 'ID', 'Country')
                 )
df1

In [ ]:
df2 = pd.DataFrame([(1, 145.0, 3000.1),
                   ( 2, 189.2, 2030.2),
                   ( 3, 129.0, 3000.0),
                   ( 4, 198.1, 4020.2)],
           columns=('DNI', 'Weight', 'Salary')
                 )
df2

In [ ]:
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html

df1.merge(df2)

In [ ]:
df1.merge(df2, left_on='ID', right_on='DNI')

In [ ]:
df2 = pd.DataFrame([(1, 145.0, 3000.1), #No 2
                   ( 3, 129.0, 3000.0), # Multiples 3
                   ( 3, 159.0, 4000.0),
                   ( 3, 109.0, 5000.0),
                   ( 4, 198.1, 4020.2),
                   ( 5, 200.0, 5000.2)], #a new one 
           columns=('DNI', 'Weight', 'Salary')
                 )
df2

In [ ]:
df1.merge(df2, left_on='ID', right_on='DNI')

In [ ]:
df1.merge(df2, left_on='ID', right_on='DNI',how="left")

In [ ]:
df1.merge(df2, left_on='ID', right_on='DNI',how="right")

In [ ]:
df1.merge(df2, left_on='ID', right_on='DNI',how="inner")

In [ ]:
df1.merge(df2,how="cross")

## <span style="color:red">Activitats</span>


### Activitat 1
Donat el següent dataframe:

In [42]:
import pandas as pd

# cream un diccionari amb les dades
data = {'País': ['Estados Unidos', 'China', 'Japón', 'Alemania', 'Reino Unido', 'India', 'Francia', 'Italia', 'Brasil', 'Canadá'],
        'PIB (billones de dólares)': [21.44, 15.54, 5.15, 4.17, 2.62, 2.61, 2.58, 1.93, 1.67, 1.65],
        'Inflación (%)': [1.8, 2.9, 0.2, 1.6, 2.2, 5.0, 1.2, 0.9, 3.7, 1.8],
        'Tasa de desempleo (%)': [3.7, 3.6, 2.9, 3.2, 3.8, 6.6, 8.1, 10.0, 13.9, 5.6],
        'Índice de competitividad': [85.6, 83.1, 82.5, 80.9, 78.9, 64.4, 78.0, 70.3, 59.5, 77.2]}

# cream el dataframe
df_economia = pd.DataFrame(data)

In [43]:
# 1. Afegeix una columna al dataframe que sigui la població de cada país segons la següent llista: `poblacion = [328.2, 1393, 126.5, 83.2, 66.9, 1380, 66.99, 60.3, 212.6, 38.1]`


In [ ]:
# 2. Afegeix una columna addicional que calculi el PIB per càpita per a cada país en dòlars dividint el PIB de cada país per la població.


In [44]:
# 3. Seleccionar els països amb una tasa d'atur major a 8.0


In [45]:
#4 . Crea un nou dataframe eliminant els països anteriorment seleccionats.

### Activitat 2

Donat els dos següents dataframes, es demana: calcular la quantitat d'ajuda rebuda per cada municipi en funció del nombre total d'habitants. 

In [46]:
import pandas as pd
import random

random.seed(0)

nombres = [f'Municipio{i}' for i in range(1, 11)] 

data_municipios = {
    'Nombre': nombres,
    'Código Postal': [random.randint(10000, 99999) for _ in range(10)],
    'Población': [random.randint(1000, 50000) for _ in range(10)]  # Añadimos un atributo aleatorio, en este caso "Población"
}

df_municipios = pd.DataFrame(data_municipios)


data_ayudas = {
    'Nombre': [random.choice(nombres) for _ in range(20)],
    'Ayuda Económica (en euros)': [random.randint(1000, 10000) for _ in range(20)],
    'Número de Beneficiarios': [random.randint(10, 100) for _ in range(20)]
}

df_ayudas = pd.DataFrame(data_ayudas)

print("Dataframe de Municipios:")
print(df_municipios)

print("\nDataframe de Ayudas:")
print(df_ayudas)


Dataframe de Municipios:
        Nombre  Código Postal  Población
0   Municipio1          60494      39232
1   Municipio2          65125      15315
2   Municipio3          15306      34075
3   Municipio4          43936      10127
4   Municipio5          77013      19470
5   Municipio6          73691      10158
6   Municipio7          63075       7214
7   Municipio8          49755      41525
8   Municipio9          72468      17417
9  Municipio10          56930      35902

Dataframe de Ayudas:
         Nombre  Ayuda Económica (en euros)  Número de Beneficiarios
0   Municipio10                        9989                       72
1    Municipio3                        1230                       23
2    Municipio5                        2528                       48
3    Municipio2                        7534                       80
4    Municipio2                        1018                       47
5    Municipio6                        9086                      100
6    Municipio8    

In [47]:
# calcular la quantitat d'ajuda rebuda per cada municipi en funció del nombre total d'habitants. 

[![License: CC BY 4.0](https://img.shields.io/badge/License-CC_BY_4.0-lightgrey.svg)](https://creativecommons.org/licenses/by/4.0/) <br/>
Authors: [Isaac Lera](https://personal.uib.cat/isaac.lera), [Miquel Miró](https://personal.uib.cat/miquel.miro) and [Biel Moyà](https://personal.uib.cat/gabriel.moya)<br/>
Institution: Universitat de les Illes Balears (UIB) <br/>